merlin/dag/executors.py

```diff
@@ -46,93 +76,152 @@ def transform(
-
-                    # Update or validate output_df dtypes
-                    for col_name, output_col_schema in node.output_schema.column_schemas.items():
-                        col_series = output_df[col_name]
-                        col_dtype = col_series.dtype
-                        is_list = is_list_dtype(col_series)
-
-                        if is_list:
-                            col_dtype = list_val_dtype(col_series)
-
-                        output_df_schema = output_col_schema.with_dtype(
-                            col_dtype, is_list=is_list, is_ragged=is_list
-                        )
-
-                        if capture_dtypes:
-                            node.output_schema.column_schemas[col_name] = output_df_schema
-                        elif len(output_df):
-                            if output_col_schema.dtype != output_df_schema.dtype:
-                                raise TypeError(
-                                    f"Dtype discrepancy detected for column {col_name}: "
-                                    f"operator {node.op.label} reported dtype "
-                                    f"`{output_col_schema.dtype}` but returned dtype "
-                                    f"`{output_df_schema.dtype}`."
-                                )
-                except Exception:
-                    LOG.exception("Failed to transform operator %s", node.op)
-                    raise
-                if output_df is None:
-                    raise RuntimeError(f"Operator {node.op} didn't return a value during transform")
+                transformed_data = self._transform_data(
+                    node, input_data, capture_dtypes=capture_dtypes
+                )
             else:
-                output_df = input_df
-
-            # Combine output across node loop iterations
+                transformed_data = input_data
 
-            # dask needs output to be in the same order defined as meta, reorder partitions here
-            # this also selects columns (handling the case of removing columns from the output using
-            # "-" overload)
-            if output is None:
-                output = output_df[node_output_cols]
-            else:
-                output = concat_columns([output, output_df[node_output_cols]])
+            output = self._combine_node_outputs(node, transformed_data, output)
 
         if additional_columns:
-            output = concat_columns([output, df[_get_unique(additional_columns)]])
+            output = concat_columns([output, data[_get_unique(additional_columns)]])
+
+        return output
+
+    def _build_input_data(self, node, data, capture_dtypes=False):
+        """
+        Recurse through the graph executing parent and dependency operators
+        to form the input dataframe for each output node
+
+        Parameters
+        ----------
+        node : Node
+            Output node of the graph to execute
+        data : DataFrameType
```

Comment:
input type here is now `Transformable`? and return type is the same or something else?